[change] NetJSON DeviceMonitoring compliance #2

purhan opened the pull request:
Closes #2, closes #64, closes #65
Resolves part of #60

purhan committed:
[change] NetJSON DeviceMonitoring compliance #2

Closes #2

purhan commented:
@pandafy @nemesisdesign In the last commit (https://github.com/openwisp/netengine/pull/62/commits/d46b5685c4bf4f63a678687fb4cde8e499f6c9cd), I've made the `to_dict` and `to_json` compliant with NetJSON DeviceMonitoring. Please take a look.
- I have not defined any new properties yet, but all the `required` fields in the schema are there.
- A lot of properties we were using are no longer required in DeviceMonitoring schema but I have not removed them yet.

pandafy commented:
> A lot of methods we were using are no longer required in DeviceMonitoring schema but I have not removed them ye

You can go ahead with this, since it will make it clear for us what methods are to be removed. We can always revert the commit if something went wrong. 

> Todo: We need to define new methods to collect as much monitoring info as possible.

instead of being exhaustive, let's start with minimal data. Check what data is being collected in [openwrt-openwisp-monitoring ](https://github.com/openwisp/openwrt-openwisp-monitoring) and add support for it here. Otherwise we may spend time collecting every possible metric many of which might not be used in OpenWISP. We can always iterate to make it better.

purhan committed:
[change] Add monitoring info to OpenWRT backend

purhan committed:
[chores] Fix failing tests

purhan committed:
[chores/qa] Improvements to tests

- Fix failing tests
- Pass more realistic values in mocks
- Remove redundant code

purhan committed:
[change] Add monitoring info to AirOS backend

pandafy reviewed (commented):
I got following errors while running tests again OpenWrt device. Is there anything that needs to be configured on the device? Is there any extra snmpd cofnigguration needed? 

```
(venv-netengine) pandafy@FRIDAY:~/openwisp/netengine$ DISABLE_MOCKS=1 TEST_SETTINGS_FILE=./test-settings.json  nose2 tests.test_snmp.test_openwrt
........EE....E.......E..
======================================================================
ERROR: test_interfaces_to_dict (tests.test_snmp.test_openwrt.TestSNMPOpenWRT)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pandafy/openwisp/netengine/tests/test_snmp/test_openwrt.py", line 80, in test_interfaces_to_dict
    self.assertIsInstance(self.device.interfaces_to_dict, list)
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 335, in interfaces_to_dict
    if_type = self.interfaces_type[i]['type']
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 281, in interfaces_type
    'type': types[self.get_value(types_oid + str(i))],
KeyError: '1'

======================================================================
ERROR: test_interfaces_type (tests.test_snmp.test_openwrt.TestSNMPOpenWRT)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pandafy/openwisp/netengine/tests/test_snmp/test_openwrt.py", line 71, in test_interfaces_type
    self.assertIsInstance(self.device.interfaces_type, list)
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 281, in interfaces_type
    'type': types[self.get_value(types_oid + str(i))],
KeyError: '1'

======================================================================
ERROR: test_netjson_compliance (tests.test_snmp.test_openwrt.TestSNMPOpenWRT)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pandafy/openwisp/netengine/tests/test_snmp/test_openwrt.py", line 120, in test_netjson_compliance
    device_dict = self.device.to_dict()
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 532, in to_dict
    'interfaces': self.interfaces_to_dict,
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 335, in interfaces_to_dict
    if_type = self.interfaces_type[i]['type']
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 281, in interfaces_type
    'type': types[self.get_value(types_oid + str(i))],
KeyError: '1'

======================================================================
ERROR: test_to_dict (tests.test_snmp.test_openwrt.TestSNMPOpenWRT)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pandafy/openwisp/netengine/tests/test_snmp/test_openwrt.py", line 113, in test_to_dict
    device_dict = self.device.to_dict()
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 532, in to_dict
    'interfaces': self.interfaces_to_dict,
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 335, in interfaces_to_dict
    if_type = self.interfaces_type[i]['type']
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/openwrt.py", line 281, in interfaces_type
    'type': types[self.get_value(types_oid + str(i))],
KeyError: '1'

----------------------------------------------------------------------
Ran 25 tests in 23.801s

FAILED (errors=4)
```

purhan commented on netengine/backends/snmp/openwrt.py:
```diff
@@ -349,60 +339,197 @@ def interfaces_to_dict(self):
+                    tzinfo=pytz.utc,
+                ).timestamp()
+            )
+        logger.warning('Invalid timestring was supplied')
```
You're right, this line won't be executed. I will remove it soon.

purhan commented on netengine/backends/snmp/airos.py:
```diff
@@ -384,6 +373,17 @@ def wireless_links(self):
+        """
+        epoch = str(self.get('1.3.6.1.4.1.41112.1.4.8.1.0')[3][0][1])
+        timestamp = int(
+            time.mktime(datetime.strptime(epoch, '%Y-%m-%d %H:%M:%S').timetuple())
```
With OpenWRT, we use a different OID, which returns an octal string and has to be converted first (for example: `\x07\xe5\x06\x0b\x06\x00\r\x00+\x00\x00`).
While here we just get a human-readable string like this: `2020-02-03 13:01:14`, so we can just extract the information using this.

pandafy replied:
Can we use `int(datetime.strptime(epoch, '%Y-%m-%d %H:%M:%S').timestamp())`

purhan replied:
Yes :+1:, we can use this

pandafy commented on netengine/backends/snmp/airos.py:
```diff
@@ -400,23 +400,74 @@ def RAM_free(self):
+        Returns an array with load average values respectively in the last
+        minute, in the last 5 minutes and in the last 15 minutes
+        """
+        one = int(self.get_value('1.3.6.1.4.1.10002.1.1.1.4.2.1.3.1'))
+        five = int(self.get_value('1.3.6.1.4.1.10002.1.1.1.4.2.1.3.2'))
+        fifteen = int(self.get_value('1.3.6.1.4.1.10002.1.1.1.4.2.1.3.3'))
```
It will be more efficient to get all three values in one request. Can you please do that?

purhan replied:
Done :+1:

purhan committed:
[chores] Some requested changes

- Swap time with datetime in local_time
- Remove unused warning
- Use a single call for collecting load information

nemesisdesign commented on netengine/backends/snmp/airos.py:
```diff
@@ -400,23 +397,75 @@ def RAM_free(self):
+        buffered = self.get_value('1.3.6.1.4.1.10002.1.1.1.1.3.0')
+        return int(buffered)
+
+    @property
```
Look at the output of the `free` command:

```
$ free
              total        used        free      shared  buff/cache   available
Mem:       16267996     7406636     2508788     1475504     6352572     7158444
Swap:       8388604       52600     8336004
```

So far I see we have total, free, then something weird, for buffered and cached we have two different values but `free` shows them as one thing.

Can you please double check and compare the output of these SNMP mibs with free?

Is there a way to know the availeble memory? See https://www.linuxatemyram.com/ for more info.

purhan replied:
@nemesisdesign is this output from an AirOS device? For me the output of `free` looks like this:
```
WA# free
             total         used         free       shared      buffers
Mem:         61960        37960        24000            0         2020
-/+ buffers:              35940        26020
Swap:            0            0            0
```

Here are the corresponding values via SNMP:
- RAM_total: `SNMPv2-SMI::enterprises.10002.1.1.1.1.1.0 = Gauge32: 61960`
- RAM_buffered: `SNMPv2-SMI::enterprises.10002.1.1.1.1.3.0 = Gauge32: 2020`
- RAM_free: `SNMPv2-SMI::enterprises.10002.1.1.1.1.2.0 = Gauge32: 24068`
- RAM_cached: `SNMPv2-SMI::enterprises.10002.1.1.1.1.4.0 = Gauge32: 0`

I'm a bit confused about the last one. On this old [thread](https://community.ui.com/questions/AP-Pro-become-disconnected-when-they-exhaust-RAM-on-3-2-1/497e0198-5727-4248-b1fa-8c13d6e41709) (and a few others in Ubiquiti forums) I found the following piece of info:
```
1.3.6.1.4.1.10002.1.1.1.1.1.0 # memTotal

1.3.6.1.4.1.10002.1.1.1.1.2.0 # memFree

1.3.6.1.4.1.10002.1.1.1.1.3.0 # memBuffer

1.3.6.1.4.1.10002.1.1.1.1.4.0 # memCached

1.3.6.1.4.1.10002.1.1.1.2.1.0 # swapTotal

1.3.6.1.4.1.10002.1.1.1.2.2.0 # swapFree
```

I haven't found anything related to available memory yet, found nothing in the standard MIBs.

PS: the output of `free` you shared looks similar to my OpenWRT device and in the openwrt backend you'll see exactly that data being collected.

nemesisdesign replied:
@purhan I see, then probably AirOS still uses an old linux kernel which doesn't support this, can you check what kernel version it uses?

What abotu OpenWRT? Do we have the information of the available memory on OpenWRT?

purhan replied:
@nemesisdesign AirOS I use for testing is running on version: `2.6.32.68`.
With OpenWRT also, I didn't find anything related to available ram even though the kernel supports it.
Here are some relevant OIDs I got from snmpwalk:
```
HOST-RESOURCES-MIB::hrStorageDescr.1 = STRING: Physical memory
HOST-RESOURCES-MIB::hrStorageDescr.3 = STRING: Virtual memory
HOST-RESOURCES-MIB::hrStorageDescr.6 = STRING: Memory buffers
HOST-RESOURCES-MIB::hrStorageDescr.7 = STRING: Cached memory
HOST-RESOURCES-MIB::hrStorageDescr.8 = STRING: Shared memory
HOST-RESOURCES-MIB::hrStorageDescr.10 = STRING: Swap space
```
I believe this means these are the only host resources we can use then?

nemesisdesign replied:
@purhan this may be the memory, did you compare these values to what you can get from `free` or `cat /proc/meminfo`?

Regarding AirOS I was not aware they used such an old kernel version. Did you already upgrade the AirOS firmware to the latest available version?

purhan replied:
> @purhan this may be the memory, did you compare these values to what you can get from `free` or `cat /proc/meminfo`?

Yes, the values match with the ones from `free`.
 
> Regarding AirOS I was not aware they used such an old kernel version. Did you already upgrade the AirOS firmware to the latest available version?

I have not upgraded yet, I will let you know more about this soon.

pandafy replied:
I compared memory results from netengine and running free command on OpenWrt and found some discrepancy. 
![Screenshot from 2021-07-05 21-34-04](https://user-images.githubusercontent.com/32094433/124497864-24349680-ddd9-11eb-976d-ba99b06b9ae3.png)

While testing I ran the free command first for control, fetched resources using netengine and then executed the free command again in hope that I will get output from netengine somewhere between these two values. 

The value of free memory was very off in my testing. We need to make sure that results from SNMP is same as from free command. Maybe we'll need to do some calculation manually.

purhan replied:
@pandafy I guess that's just a drawback of SNMP? I did some research on this and found some [discussions](https://serverfault.com/questions/640459/snmp-memory-values-do-not-match-free) but no luck

purhan replied:
@pandafy @nemesisdesign I upgraded the device to the latest firmware (v8.7.4). Output from free command appears to be the same:
```
WA# free
             total         used         free       shared      buffers
Mem:         61960        45268        16692            0         4128
-/+ buffers:              41140        20820
Swap:            0            0            0
```
Kernel version is the same for this (i.e. 2.6.32.68)
I don't see a difference in the OIDs.
Tested netengine and the results are similar, so no noticeable difference yet.

purhan committed:
[chores] Fix neighbor information when mac address is not found

pandafy commented on netengine/backends/snmp/openwrt.py:
```diff
@@ -349,60 +339,196 @@ def interfaces_to_dict(self):
+            '7': 'INCOMPLETE',
+        }
+
+        # TODO: find a way to extract IP address from the OID
```
I spent some time on this and this is what i found for OpenWrt device. 

Following is output for my laptop as neighbour of the router. I replaced my MAC with `FF FF FF FF FF FF`.

```
$ snmpwalk -v2c -c public  192.168.1.1 1.3.6.1.2.1.4.35.1.4
iso.3.6.1.2.1.4.35.1.4.5.1.4.192.168.1.136 = Hex-STRING: FF FF FF FF FF FF
iso.3.6.1.2.1.4.35.1.4.5.2.16.254.128.0.0.0.0.0.0.35.113.163.173.241.78.112.113 = Hex-STRING: FF FF FF FF FF FF
iso.3.6.1.2.1.4.35.1.4.5.2.16.254.128.0.0.0.0.0.0.77.92.143.78.43.2.155.123 = Hex-STRING: FF FF FF FF FF FF
```

You need to pay attention for the OID here. For IPv4 addresses the 12th index is `4` and for IPv6 address 12th index is `16`. 

The IPv6 addresses are represented in decimal octets here. You will have to convert them to hexadecimals. Check if there's an existing library which can do this to make our work simple. 

Check this pseudocode for example, improve it wherever required:

```
for neighbor in neighbors:
    oid = neighbour[0][0].getOid()
    if oid[12] == 4:
        ip = oid[13:]
    else:
        ip = convert_decimal_octets_to_hexademical_blocks(oid[13:])
```

purhan replied:
Done :+1:

pandafy reviewed (changes requested):
Great work Purhan! Please see my comments below.

purhan committed:
[change] Add IP to neighbor information

purhan commented on netengine/backends/snmp/openwrt.py:
```diff
@@ -349,60 +339,196 @@ def interfaces_to_dict(self):
+        }
+
+        # TODO: find a way to extract IP address from the OID
+        neighbors = self.next('1.3.6.1.2.1.4.35.1.4')[3]
+        neighbor_states = self.next('1.3.6.1.2.1.4.35.1.7')[3]
```
It may be possible, I will look into it

pandafy commented on netengine/backends/snmp/openwrt.py:
```diff
@@ -349,60 +339,206 @@ def interfaces_to_dict(self):
+        """
+        returns the free RAM of the device
+        """
+        RAM_used = int(self.get_value('1.3.6.1.2.1.25.2.3.1.6.1'))
```
If we are fetching `used` memory then why we are not showing it?

purhan replied:
Only because the [devicemonitoring](https://netjson.org/rfc.html#name-devicemonitoring) spec didn't mention it. If we use it somewhere then yes we can show it :+1:

pandafy commented on netengine/backends/snmp/base.py:
```diff
@@ -57,6 +59,20 @@ def _octet_to_mac(self, octet_mac):
             )
         return mac_address
 
+    def _ascii_blocks_to_ipv6(self, ascii_string):
```
Are you sure this is ASCII representation of IPv6? I am not sure about this.

purhan replied:
Yes, these are the ascii values converted into hexadecimal by this function. You can look at [this chart](https://www.garykessler.net/library/ascii.html) for example.
For example if an ip address is `fe80::....` then we should see something like `254.128....` in the oid

pandafy commented on netengine/backends/snmp/base.py:
```diff
@@ -57,6 +59,20 @@ def _octet_to_mac(self, octet_mac):
+        converts an ascii representation into ipv6 address
+        """
+        blocks = ascii_string.split('.')
+        for b in range(len(blocks)):
+            blocks[b] = format(int(blocks[b]), '02x')
+        res = netaddr.IPAddress(
+            ':'.join(
+                [''.join(blocks[slice(i, i + 2)]) for i in range(0, len(blocks), 2)]
+            )
+        )
+        return res
```
Can you please add comments how this function achieves the conversion?

pandafy reviewed (commented):
While testing it was found that `to_dict` methods are taking too long to fetch the data since it is making multiple queries. To make this method quicker, we can do a single query to device for a complete SNMP Walk and then process the data as needed locally by netengine. 


I have been thinking about how we can bring this change and here is one of my idea. 

Instead of changing complete implementation we can update the signature of current methods to accept another optional parameter `snmpdump` which will be the output from snmpwalk. The methods will locally fetch the required OIDs from snmpdump and then the existing code for further processing can be used. 

Here's a pseudocode for example:

```
def local_time(self, snmpdump=None):
    if snmpdump:
        epoch = get_local_time_from_snmp_dump
    else:
        epoch = str(self.get('1.3.6.1.4.1.41112.1.4.8.1.0')[3][0][1])
     timestamp = int(datetime.strptime(epoch, '%Y-%m-%d %H:%M:%S').timestamp())
     return timestamp
```

purhan committed:
[chores] Add comment

purhan committed:
[change] Add more interfaces, Add RAM_used

purhan committed:
[change] Use one query for neighbors and their states

nemesifier reviewed (commented):
Thanks @purhan, with that fix I can now extract the info without errors.

For sure we'll have find a way to get all the info at once to resolve the performance issue when generating the JSON output, in the current form it would not be usable.

In the output of type I see `ethernet Csmacd`, but to be compliant with the NetJSON spec it should be just `ethernet`.

One thing I noticed is that wifi interfaces are also detected as ethernet and I don't see any wireless info is that the max we can do?
Looking at the snmpd package for OpenWRT, I do see a patch that seems to add support for WiFi information: https://github.com/openwrt/packages/blob/38f01ad2c9738c6c5a5dbda85019bb9cf5bc10f8/net/net-snmp/patches/750-ieee802dot11.patch.

[I just asked on the OpenWRT forum](https://forum.openwrt.org/t/snmpd-wifi-info/100921) if the know anything about it.

pandafy commented on netengine/backends/snmp/utils.py:
```diff
@@ -0,0 +1,236 @@
+ifTypes = {
```
Add reference to the source location.

purhan committed:
[chores] Fix failing build

purhan committed:
[change] Implement SNMPDUMP

nemesifier reviewed (changes requested):
from snmpwalk:

iso.3.6.1.2.1.1.1.0 = STRING: "Linux Capoano-Test-01 4.14.229 #0 Fri Apr 9 13:43:38 2021 mips"

From this informatio we can find out the device name, which would be nice to have in the output. The device name is what comes after Linux, but before the kernel version.

Then onto the IP addresses of each interface, I see they're available in the snmpwalk output but I can't get this info from the JSON dict:

```
iso.3.6.1.2.1.4.20.1.1.10.10.0.48 = IpAddress: 10.10.0.48
iso.3.6.1.2.1.4.20.1.1.127.0.0.1 = IpAddress: 127.0.0.1
iso.3.6.1.2.1.4.20.1.1.192.168.1.59 = IpAddress: 192.168.1.59
iso.3.6.1.2.1.4.20.1.1.192.168.254.1 = IpAddress: 192.168.254.1
```

Can you make sure this info is returned? It's very useful to display this info in the status tab.

pandafy commented on netengine/backends/snmp/openwrt.py:
```diff
@@ -545,8 +545,10 @@ def neighbors(self):
 
-    def to_dict(self):
-        return self._dict(
+    def to_dict(self, snmpdump=False):
```
I think `to_dict` should always be run with `snmpdump=True`

purhan committed:
[change] Add device name, ip to interfaces

purhan committed:
[change] Make interface types NetJSON compliant

nemesifier reviewed (changes requested):
Good progress @purhan :+1: 

Now we need to start digging deeper into OpenWRT/SNMPd in order to add more information which is there but we haven't been looking at and clean up things a bit.

Here's a few issues I see:

- ip address format is not valid NetJSON, please double check the spec and the output returned by the netjson-monitoring scripts
- total memory got via netjson-monitoring scripts: `127467520`, via netengine I get `124480`, looks like the value we get from SNMPd is not in bytes but KB? Make sure the size used is the same, multiply if necessary.
- given the issue above, please double check the interface traffic counters, size must be consistent, we can't use bytes somewhere and kilobytes elsewhere
- load average is available with `1.3.6.1.4.1.2021.10`, please add this info to the output because it's useful
- memory: with oid `1.3.6.1.4.1.2021.4` I can get values that are more similar to those found in /proc/meminfo:

```
"memory": {
    "total": 127467520,
    "free": 89944064,
    "shared": 135168
    "buffered": 4501504,
    "cached": 13062144,
    "available": 74657792,
},
```

iso.3.6.1.4.1.2021.4.5.0 = INTEGER: 124480   (memTotalReal, 124480 * 1024 = 127467520, same value retrieved with netjson_monitoring in "total")
iso.3.6.1.4.1.2021.4.6.0 = INTEGER: 87956  (memAvailReal, 87956 * 1024 = 90066944, this is closer to "free" rather than "available")
iso.3.6.1.4.1.2021.4.11.0 = INTEGER: 87956 (memTotalFree, same as before)
iso.3.6.1.4.1.2021.4.12.0 = INTEGER: 16000 (memMinimumSwap)
iso.3.6.1.4.1.2021.4.13.0 = INTEGER: 132 (memShared, 132 * 1024 = 135168, same value in "shared" from netjson_monitoring)
iso.3.6.1.4.1.2021.4.14.0 = INTEGER: 4396 (memBuffer, 439 * 1024 = 4501504, same as "buffered" above)
iso.3.6.1.4.1.2021.4.15.0 = INTEGER: 12756 (memCached, 12756 * 1024 = 13062144, same value as "cached above")

Therefore for memory I think we should use this MIB.

- please use the MIB with oid `1.2.840.10036` to determine which interfaces are of type WiFi
- please double check the same also on AirOS

I left many more pointers in our chat, please read those.

CC: @pandafy.

purhan committed:
[change] Change RAM OIDs, add load info, add wireless interfaces

pandafy commented on netengine/backends/snmp/base.py:
```diff
@@ -94,7 +95,26 @@ def _oid(self, oid):
 
-    def get(self, oid):
+    def walk(self, oid):
+        result = dict()
```
Why are you not directly creating Trie?

purhan replied:
This was a mistake, thanks I've changed it.

purhan committed:
[change] Use Trie instead of Dict in walk function

nemesifier reviewed (changes requested):
Ok @purhan it's time to improve the docs a bit more, let's document the main methods like `validate`, `to_dict`, `to_json`, then we should have a page for OpenWrt and a page for AirOS, in each of the two we should have a list which explains the data returned with the to_json method (eg: interfaces, system, neighbours, etc), we should mention the format is inspired by NeJSON DeviceMonitoring.

I think we can point out that the internal methods that are not documented may be changed in future versions, we could also flag all the other methods as private if you think it's a good idea.

purhan committed:
[docs] Update docs for AirOS and OpenWRT backends

purhan committed:
[chores] Move OpenWRT interface type to match AirOS

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+the tree gains more depth.
+Obviously, by getting the smallest MIB which is "1" or simply " . " one can get all the tree.
+
+The base SNMP backend contains the following methods (some internal methods are not documented and are subject to change in the future):
```
I think we can avoid explaining what SNMP is, we assume peopel using this library will know.

We can however include information about how all the data is retrieved and give a few hints on how to use the base backend to create new SNMP backends for other systems (eg: indicate the python path for the base backend and explain anything important regarding it).

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+
+The SNMP backend provides support for 2 firmwares:
+ * AirOS
+ * OpenWRT
```
please list OpenWrt first. Please also link their respective pages

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+| **agent**     | Agent string for the SNMP connection                                |
++---------------+---------------------------------------------------------------------+
+| **port**      | Port for the SNMP connection. Default value is `161`                |
++---------------+---------------------------------------------------------------------+
```
Are these parameters all mandatory?
Is there any default value for any parameter?
Let's make this explicit.
There's no need to write this table, sphinx has automations we can use to autodocument these things.
Please use it!
Refer to the netjsonconfig documentation for examples:
http://netjsonconfig.openwisp.org/en/latest/backends/openwrt.html#initialization

I mean this:

![Screenshot from 2021-08-19 17-31-32](https://user-images.githubusercontent.com/841044/130152817-1eb02414-a2d1-4d49-9b33-fe8e30ab2c5c.png)

See the source for more info: https://raw.githubusercontent.com/openwisp/netjsonconfig/master/docs/source/backends/openwrt.rst

The doc text is embedded in a docstring in the method itself.

This is a good practice, the docstring is also shown in some rich shells (bpython should show it, not sure about ipython).

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+| **to_json**  | Calls the `to_dict` method and returns a JSON string of the dict                                                                         |
++--------------+------------------------------------------------------------------------------------------------------------------------------------------+
+| **validate** | Checks if connection with the device is working and raises `NetengineError` in case something is wrong                                   |
++--------------+------------------------------------------------------------------------------------------------------------------------------------------+
```
I suggest to do the same here and drop the table syntax which is painful to maintain

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+*****
+AirOS
+*****
+
```
as in netjsonconfig, I suggest to make this a dedicated page, which should point to the base snmp page for the general concepts.

Some information regarding methods can be duplicated (no problem with this if we use autodoc).

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+With AirOS, Netengine is able to collect the following information which is returned in the
+`NetJSON Devicemonitoring <https://netjson.org/rfc.html#name-devicemonitoring>`_ format:
+
++------------+------------------------------------------------------------------------------------------------------------+
```
please let's switch from table sytnax which is painful to maintain to simple headings and list after headings.

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+With OpenWRT, Netengine is able to collect the following information which is returned in the
+`NetJSON Devicemonitoring <https://netjson.org/rfc.html#name-devicemonitoring>`_ format:
+
++------------+----------------------------------------------------------+
```
please let's switch from table sytnax which is painful to maintain to simple headings and list after headings.

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+ device.next("1.3.6")
+
+Otherwise, if you want simply a value of the tree just type::
+ device.get_value("oid_you_want_to_ask_for")
```
these methods are not documented above so why are we showing them in the example?

I guess we can document `next` and `get_value`, but I would avoid documenting `uptime_tuple` because it would not be consistent with advising to use `to_netjson`.

nemesisdesign commented on docs/source/topics/backends/snmp.rst:
```diff
@@ -0,0 +1,149 @@
+OpenWRT example
+===============
+
+The same instructions typed above can be applied to OpenWRT itself, just remember to import the correct firmware by typing::
```
it's not a firmware, it's a backend.

nemesifier reviewed (changes requested):
Thank you @purhan, I see good progress here but there's room for improvement, we should put effort in improving this further.

I highly advise to make it explicit that we recommend using only the `to_netjson` method to retrieve information and that all the other methods right now are not stable, I ask you to please flag all the other methods as private by prepending an underscore, eg: `method_name()` becomes `_method_name`. This way we are signaling to others who may want to use this code that we reserve the freedom to change the internal API freely if we need.
We should do this because the library is still in early stages and it's generally recommended to be conservative in which methods are declared public and can be used by other developers, so that only the minimum needed has to be maintained with backward compatibility in mind (which requires more work) and the rest can be freely changed.
I am pretty sure that over the next iterations we'll be changing a lot of internal code so it's better to prepare this right now.

Please dedicate one page to OpenWrt and one page to AirOS.
Please list OpenWrt before AirOS.

Please refer to the netjsonconfig documentation http://netjsonconfig.openwisp.org/en/latest/ for insipiration on how to write documentation which is informative and consistent with the rest of OpenWISP, this documentation is built with ReStructuredText and Python Sphinx just like this one, you can find the source here: https://github.com/openwisp/netjsonconfig/tree/master/docs/source.

I'll be referring to more specific examples below.

nemesifier reviewed (changes requested):
Missing things we found out in AirOS during our last call:

- device name
- device model (Ubiquiti LiteBeam ...)
- operating system (Linux version (withou date) + firmware version)
- please check what the mac addresses are and if we can display them (if these mac addresses are coming from the interfaces, we should include them)

purhan committed:
[change] Add CPU_count to airOS

nemesifier committed:
[qa] Reformatted code and docs

nemesifier committed:
Merge branch 'master' into netjson-compliance

nemesifier committed:
[fix] Added back jsonschema

nemesifier committed:
[chores] Added device model, OS

nemesifier committed:
[chores] Fixed bugs uncovered by Kilobot

nemesifier commented on tests/test_dummy.py:
```diff
@@ -24,6 +24,6 @@ def test_get_manufacturer(self):
 
     def test_to_json(self):
-        json_string = self.dummy.to_json()
+        json_string = self.dummy.to_json(autowalk=False)
```
`Dummy.to_dict()` accepts arbitrary keyword arguments and does not use `autowalk`, so `to_json()` has identical behavior with the default and with `autowalk=False`. A default-path assertion would not cover any additional behavior. No change needed.

nemesifier commented on netengine/backends/snmp/base.py:
```diff
@@ -70,80 +71,83 @@ async def _walk(self, oid):
+                result[str(var_bind[0])] = [None, None, None, [var_bind]]
+        return result
+
+    def get(self, oid, snmpdump=None):
+        """Execute an SNMP GET request or read its value from an SNMP dump."""
+        oid = self._oid(oid)
+        if snmpdump is not None:
+            return snmpdump.get(oid, [None, None, None, [[None, ""]]])
+        logger.info("SNMP GET %s", oid)
+        return asyncio.run(self._command(get_cmd, oid))
```
This is not a demonstrated OpenWISP integration bug. NetEngine intentionally treats missing OIDs in sparse or partial dumps as empty values because the base SNMP layer cannot determine whether an OID is unsupported, unavailable on a firmware, intentionally omitted, or outside the collected subtree. Raising globally would break existing sparse-MIB probing such as `interfaces_speed()`.

The concrete issue is that default serialization currently walks incomplete roots, so required OIDs can be omitted from its own dump. We will fix those autowalk roots. We will keep the dump-miss fallback and address backend-specific numeric handling only if a real OpenWISP partial-dump use case demonstrates a failure.

nemesifier committed:
[fix] Addressed @coderabbitai reviews

nemesifier committed:
[fix] Fixed more nasty bugs

nemesifier committed:
[fix] Renamed memory "cached" (bogus) to "cache" (correct)

nemesifier committed:
[chores] Follow up with coderabbit

nemesifier commented:
@coderabbitai full review

nemesifier commented on tests/test_snmp/test_airos.py:
```diff
@@ -34,19 +38,21 @@ def setUp(self):
-        with patch(
-            "netengine.backends.snmp.base.get_cmd", side_effect=NoSuchObjectError
-        ):
-            with self.assertRaises(NoSuchObjectError):
-                self.device.get_value(".")
+        with self.assertRaises(NoSuchObjectError):
+            self.device.get_value(".")
```
Verified locally: `get_value(".")` raises `NoSuchObjectError` before an SNMP request is dispatched. The focused test completes in 0.135 seconds with the command mock removed, so it deterministically tests PySNMP OID validation without network I/O.

nemesifier commented on tests/static/test-openwrt-snmp-oid.json:
```diff
@@ -41,9 +41,49 @@
+  "1.3.6.1.2.1.25.2.3.1.6.8": "88",
+  "1.3.6.1.2.1.25.2.3.1.6.10": "0",
+  "1.3.6.1.2.1.25.2.3.1.5.10": "0",
+  "1.3.6.1.2.1.25.1.2.0": {
+    "type": "bytes",
+    "value": "\\x07\\xe5\\x06\\x0b\\x06\\x00\r\\x00+\\x00\\x00"
+  },
```
The fixture contains the JSON escape sequence `\\r`, not a raw carriage-return byte: its source bytes are `5c 72`. JSON decoding yields the intended byte before `codecs.escape_decode`, so line-ending normalization cannot alter this fixture. The proposed representation is equivalent but unnecessary.

nemesifier committed:
[chores] Addressed @coderabbitai review comments

nemesifier committed:
Merge remote-tracking branch 'origin/master' into netjson-compliance

# Conflicts:
#	.github/workflows/pypi.yml

nemesifier committed:
[chores] Addressed @coderabbitai review comments

nemesifier commented:
@coderabbitai review

nemesifier committed:
[chores] Addressed @coderabbitai review comments

nemesifier commented:
@coderabbitai resolve

nemesifier commented on .github/workflows/pypi.yml:
```diff
@@ -17,16 +17,16 @@ jobs:
+      - name: Build package
+        run: python -m build
+      - name: Publish package distributions to PyPI
+        uses: pypa/gh-action-pypi-publish@v1.14.2
```
skip this please

nemesifier committed:
[chores] Addressed @coderabbitai review comments

nemesifier commented:
@coderabbitai review

nemesifier committed:
[chores] Further cleanup

nemesifier commented:
@coderabbitai resolve
